### PR TITLE
Issue/1403 `create-secrets` utility for helping users to create k8s secrets easier

### PR DESCRIPTION
**Pull request by t83714**
### How to use:
Go to Magda project root directory & run:
```
`yarn bin`/create-secrets
```
You can type:
```
`yarn bin`/create-secrets --help
```
to print the help screen.

#### Save config data to a file
```
`yarn bin`/create-secrets -P > config.json
```
#### Create secrets using config from STDIN or a file

```
cat config.json | `yarn bin`/create-secrets -E -
```
or
```
CONFIG_DATA=`config.json`
echo $CONFIG_DATA | `yarn bin`/create-secrets -E -
```
or
```
 `yarn bin`/create-secrets -E ./config.json
```
#### You can choose to overwrite some settings by ENV variable
e.g. You can create secrets using config data `$CONFIG_DATA` but only overwrite `SMTP_SECRET_PASSWORD`.
```
export SMTP_SECRET_PASSWORD="qazwsx"
echo $CONFIG_DATA | `yarn bin`/create-secrets -E -
```
A list of all available ENV variables can be found from help screen.

Please note: If you answered `NO` to question `Do you want to allow environment variables to overwrite current settings at runtime?`, no ENV variables will be used (except `$CI_COMMIT_REF_SLUG` & `$CI_JOB_TOKEN` depends on your answers to other questions).

#### Other Notes for test:

- File picker will only list all json files from current directory to 2 level deeper.

### Checklist

-   [x] Unit tests aren't applicable
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (or there is no issue) and dragged it to the _Pull Request_ column
-   [x] There are sufficient comments for my code to be understandable - and I realise reviewers will pull me up on it if not!

### CODE REVIEW

File: scripts/create-secrets/askQuestions.js [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+            {
+                name: "NO (No settings will be override)",
+                value: false
+            }
+        ]
+    }
+];
+
+function onlyAvailableForGoogleCloud(anwsers) {
+    return anwsers["deploy-to-google-cloud"];
```
Comment (AlexGilleran) [inline scripts/create-secrets/askQuestions.js]: answers

File: scripts/create-secrets/askQuestions.js [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+}
+
+function onlyWhenQuestion(name, value) {
+    return anwsers => {
+        return anwsers[name] === value;
```
Comment (AlexGilleran) [inline scripts/create-secrets/askQuestions.js]: answers

File: scripts/create-secrets/index.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            }
+            config.all = data;
+            return k8sExecution(config, true);
+        })
+        .catch(function(error) {
+            hasError = true;
+            console.log(chalk.red(`Failed to create secrets: ${error}`));
+        })
+        .then(function() {
+            //--- cann't use finally
```
Comment (AlexGilleran) [inline scripts/create-secrets/index.js]: can't

File: scripts/create-secrets/k8sExecution.js [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    "combined-db-client",
+    "discussions-db",
+    "discussions-db-client",
+    "registry-db",
+    "registry-db-client",
+    "session-db",
+    "session-db-client"
+];
+
+let env = process.env;
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: Is it possible to keep this off the global namespace and use whatever kind of env is needed at the time? Having it here as a `let` and setting it as a side effect means when I read the code I'm never sure what env currently is.

File: scripts/create-secrets/k8sExecution.js [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        configData["manual-db-passwords"]["password"] &&
+        !env[settingNameToEnvVarName("db-passwords")]
+    ) {
+        configData["db-passwords"] =
+            configData["manual-db-passwords"]["password"];
+    }
+    return configData;
+}
+
+function validKubectl() {
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: Can we change this to `checkIfKubectlValid` or something like that?

File: scripts/create-secrets/k8sExecution.js [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        return doK8sExecution(config, shouldNotAsk);
+    } catch (e) {
+        return Promise.reject(e);
+    }
+}
+
+function doK8sExecution(config, shouldNotAsk = false) {
+    env = getEnvByClusterType(config);
+    let configData = Object.assign({}, config.all);
+    const ifAllowEnvVarOverride = configData["allow-env-override-settings"];
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: Could this be `allowEnvVarOverride` ? So we don't end up with `if (if`

File: scripts/create-secrets/k8sExecution.js [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+    if (ifAllowEnvVarOverride) {
+        configData = overrideSettingWithEnvVars(configData);
+    }
+    validKubectl();
+    let p = Promise.resolve().then(function() {
+        configData["cluster-namespace"] = trim(configData["cluster-namespace"]);
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: This stuff doesn't really need to be in `then` right?

File: scripts/create-secrets/k8sExecution.js [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+
+function doK8sExecution(config, shouldNotAsk = false) {
+    env = getEnvByClusterType(config);
+    let configData = Object.assign({}, config.all);
+    const ifAllowEnvVarOverride = configData["allow-env-override-settings"];
+    if (ifAllowEnvVarOverride) {
+        configData = overrideSettingWithEnvVars(configData);
+    }
+    validKubectl();
+    let p = Promise.resolve().then(function() {
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: can we name `p` `promise` or something? Not much difference but it makes it a bit easier to read.

File: scripts/create-secrets/k8sExecution.js [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+                `Namespace ${
+                    configData["cluster-namespace"]
+                } doesn't exist. Please create and try again.`
+            );
+        }
+        p = p
+            .then(
+                askIfCreateNamespace.bind(null, configData["cluster-namespace"])
+            )
+            .then(function(ifCreate) {
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: Can we rename `ifCreate` to `shouldCreateNamespace` or something like that?

File: scripts/create-secrets/k8sExecution.js [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+                if (!ifCreate) {
+                    console.log(
+                        chalk.yellow(
+                            `You need to create namespace \`${
+                                configData["cluster-namespace"]
+                            }\` before try again.`
+                        )
+                    );
+                    process.exit();
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: `process.exit(1)` if it's an error 👍

File: scripts/create-secrets/k8sExecution.js [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+                } else {
+                    createNamespace(configData["cluster-namespace"]);
+                }
+            });
+    }
+    return p.then(function() {
+        const namespace = configData["cluster-namespace"];
+        if (configData["use-cloudsql-instance-credentials"] === true) {
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: Can we put blank lines around these ifs? It kind blurs into a big wall of text for me.

File: scripts/create-secrets/k8sExecution.js [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+                configData["storage-account-credentials"]
+            );
+        }
+        if (configData["use-smtp-secret"] === true) {
+            createSecret(namespace, "smtp-secret", {
+                username: configData["smtp-secret-username"],
+                password: configData["smtp-secret-password"]
+            });
+        }
+        (function() {
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: Why is this in an immediately executing function?

File: scripts/create-secrets/k8sExecution.js [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+            }
+            if (configData["use-oauth-secrets-facebook"]) {
+                data["facebook-client-secret"] =
+                    configData["oauth-secrets-facebook"];
+            }
+            createSecret(namespace, "oauth-secrets", data);
+        }
+
+        (function() {
+            const data = {};
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: Once again, why the function wrapper?

File: scripts/create-secrets/k8sExecution.js [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+}
+
+function createNamespace(namespace) {
+    childProcess.execSync(`kubectl create namespace ${namespace}`, {
+        stdio: "inherit",
+        env: env
+    });
+}
+
+function getTplObj(name, namespace) {
```
Comment (AlexGilleran) [inline scripts/create-secrets/k8sExecution.js]: I assume this is to get a template object? Can we call it `buildTemplateObject` ?

File: scripts/create-secrets/prompts/inquirer-fuzzy-path.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+const Choices = require("inquirer/lib/objects/choices");
+const InquirerAutocomplete = require("inquirer-autocomplete-prompt");
+const stripAnsi = require("strip-ansi");
+const style = require("ansi-styles");
+const fuzzy = require("fuzzy");
+
+const readdir_ = util.promisify(fs.readdir);
+
+class InquirerFuzzyPath extends InquirerAutocomplete {
```
Comment (AlexGilleran) [inline scripts/create-secrets/prompts/inquirer-fuzzy-path.js]: Could we put some JsDoc on this? I'm not sure what `InquirerFuzzyPath` is supposed to do just from the name.

File: scripts/create-secrets/prompts/inquirer-fuzzy-path.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    constructor(question, rl, answers) {
+        const rootPath = question.rootPath || ".";
+        const pathFilter = question.pathFilter || (() => true);
+        const _question = Object.assign({}, question, {
+            source: (_, pattern) => getPaths(rootPath, pattern, pathFilter)
+        });
+        super(_question, rl, answers);
+    }
+    search(searchTerm) {
```
Comment (AlexGilleran) [inline scripts/create-secrets/prompts/inquirer-fuzzy-path.js]: can we put blank lines between the methods here?

File: scripts/create-secrets/prompts/inquirer-fuzzy-path.js [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+                    validationResult || "Enter something, tab to autocomplete!"
+                );
+                return;
+            }
+        }
+        super.onSubmit(stripAnsi(line));
+    }
+}
+
+const maxScanDepth = 2;
```
Comment (AlexGilleran) [inline scripts/create-secrets/prompts/inquirer-fuzzy-path.js]: If this is a const can it go at the top of the file?

File: scripts/create-secrets/prompts/list-with-transformer.js [hunk 18]
```diff
@@ -0,0 +1,94 @@
+const List = require("inquirer/lib/prompts/list");
+const chalk = require("chalk");
+const figures = require("figures");
+const cliCursor = require("cli-cursor");
+
+class listWithTransformer extends List {
```
Comment (AlexGilleran) [inline scripts/create-secrets/prompts/list-with-transformer.js]: If this is a class it should start with a capital - i.e. `ListWithTransformer`

File: doc/deploying-for-production-on-gke.md [hunk 19]
```diff
@@ -2,44 +2,36 @@ NOTE: Work in progress... this probably won't work 100% for you. But it's a star
...
 1.  Connect to the cluster you want to deploy on with kubectl
 2.  Install helm https://github.com/kubernetes/helm/blob/master/docs/install.md
-3.  Create secrets:
+3.  Create necessary k8s secrets:
+
+Please run `create-secrets` tool:
 
 ```bash
-deploy/helm/create-auth-secrets.sh
+`yarn bin`/create-secrets
```
[the remark this answers is not in the document]
Comment (t83714) [reply doc/deploying-for-production-on-gke.md]: it's for computers that haven't setup `path` properly for yarn.
I guess we probably need it? --- correct me if I am wrong
yarn bin will only return the path to `.bin` directory?
So, `yarn bin create-secrets` will only output the path like `magda4/node_modules/.bin/create-secrets`?
You have to include ` like
```
`yarn bin`/create-secrets
```
or 
```
`yarn bin create-secrets`
```
to run it?

File: doc/deploying-for-production-on-gke.md [hunk 20]
```diff
@@ -2,44 +2,36 @@ NOTE: Work in progress... this probably won't work 100% for you. But it's a star
...
+-   **useCloudSql**: Do you want to use Google Cloud SQL for the databases? This should be used with `useCombinedDb=false`
+-   **externalUrl**: What's the external url that you'll be deploying the website on? This is used for OAuth2 callback URLs
+-   **indexer.elasticsearch.useGcsSnapshots**: Do you want to have elasticsearch snapshot to GCS? If so you'll need to create a GCS bucket for it, and set that in gcsSnapshotBucket. You'll also need to make sure your GKE cluster has access to GCS when you create it.
+-   **gateway.loadBalancerIP**: What's the external IP you want to use?
+-   **gateway.auth.x**: Put the ids of your google/facebook apps for OAuth if you have them.
+-   **feedback-api.gitHubIssuesUrl**: Put the API URL of your (private) GitHub repo where feedback issues will be created, e.g. `https://api.github.com/repos/TerriaJS/Magda-Feedback/issues`. You also need to create secret called `access-tokens` with a key `github-for-feedback` containing a personal access token with permissions to create issues in the private repo (this secret is currently not covered by `create-secrets` tool in step 2).
```
[the remark this answers is not in the document]
Comment (t83714) [reply doc/deploying-for-production-on-gke.md]: Will delete this~ 👍

File: scripts/create-secrets/askQuestions.js [hunk 21]
```diff
@@ -0,0 +1,664 @@
...
+            return true;
+        },
+        validate: validJsonFileExist,
+        filter: getJsonFileContent
+    },
+    {
+        type: "list",
+        name: "use-smtp-secret",
+        message:
+            "Do you need to access STMP service for sending data request email?",
```
[the remark this answers is not in the document]
Comment (t83714) [reply scripts/create-secrets/askQuestions.js]: Thanks a lot~ Will correct this~

File: scripts/create-secrets/askQuestions.js (hunk 1 above)
Comment (t83714) [reply scripts/create-secrets/askQuestions.js]: 👍

File: scripts/create-secrets/askQuestions.js (hunk 2 above)
Comment (t83714) [reply scripts/create-secrets/askQuestions.js]: 👍 oops~

File: scripts/create-secrets/index.js (hunk 3 above)
Comment (t83714) [reply scripts/create-secrets/index.js]: Will correct this~ thank you~

File: doc/deploying-for-production-on-gke.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (AlexGilleran) [reply doc/deploying-for-production-on-gke.md]: Oh right that's cool then, I thought it was just leftover markdown for the code snippet.

Maybe we can put it in scripts so the command is just `yarn run create-secrets` ?

File: scripts/create-secrets/k8sExecution.js (hunk 4 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: Will pass it as a function parameter to all functions need it~ 👍

File: scripts/create-secrets/k8sExecution.js (hunk 6 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: Good point~ Will update this~😄

File: scripts/create-secrets/k8sExecution.js (hunk 7 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: The reason I put the code in a then chain is I want to make sure exceptions thrown will be captured at the end of the then chain (outside `doK8sExecution` function).
I think I should also include `validKubectl();` in `then` (missed it). What do you think? @AlexGilleran

File: scripts/create-secrets/k8sExecution.js (hunk 9 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: 👍

File: scripts/create-secrets/k8sExecution.js (hunk 8 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: 👍

File: scripts/create-secrets/k8sExecution.js (hunk 10 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: 👍

File: scripts/create-secrets/k8sExecution.js (hunk 11 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: 👍 I am also the fan of bigger space between code~

File: scripts/create-secrets/k8sExecution.js (hunk 12 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: Only want to create a new scope so I don't need to worry about any temporary variables I defined.
In this case, `const data`

File: scripts/create-secrets/k8sExecution.js (hunk 7 above)
Comment (AlexGilleran) [reply scripts/create-secrets/k8sExecution.js]: Oh right - ok makes sense then 👍

Comment (AlexGilleran) [reply scripts/create-secrets/k8sExecution.js]: Can you put that in a comment?

File: scripts/create-secrets/k8sExecution.js (hunk 13 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: As above, just want to create a new scope

File: scripts/create-secrets/k8sExecution.js (hunk 14 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: 👍

File: scripts/create-secrets/prompts/inquirer-fuzzy-path.js (hunk 15 above)
Comment (t83714) [reply scripts/create-secrets/prompts/inquirer-fuzzy-path.js]: Good idea~ Will put some description~ This class is a plugin to work as file picker.

File: scripts/create-secrets/prompts/inquirer-fuzzy-path.js (hunk 16 above)
Comment (t83714) [reply scripts/create-secrets/prompts/inquirer-fuzzy-path.js]: 👍

File: scripts/create-secrets/prompts/inquirer-fuzzy-path.js (hunk 17 above)
Comment (t83714) [reply scripts/create-secrets/prompts/inquirer-fuzzy-path.js]: 👍

File: scripts/create-secrets/k8sExecution.js (hunk 12 above)
Comment (AlexGilleran) [reply scripts/create-secrets/k8sExecution.js]: If it makes sense to have it in its own function, you might as well create an actual named function right? Like `function createDbPasswords(dbPasswordNames) {`

File: doc/deploying-for-production-on-gke.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (t83714) [reply doc/deploying-for-production-on-gke.md]: @AlexGilleran 
Yeah~ that's easier to use. Have added to scripts and updated documents accordingly~

**Comment by AlexGilleran** [comment]
OK had a test, works really good. My only comment is that instead of the CI_COMMIT_REF_SLUG thing, that question should be "specify a namespace or use default namespace" and then that should be overridable by an env variable... $CI_COMMIT_REF_SLUG is a specific Gitlab thing, I don't think it really needs to be in our tooling, especially use facing tooling.

File: scripts/create-secrets/prompts/list-with-transformer.js (hunk 18 above)
Comment (t83714) [reply scripts/create-secrets/prompts/list-with-transformer.js]: updated this~

File: scripts/create-secrets/k8sExecution.js (hunk 12 above)
Comment (t83714) [reply scripts/create-secrets/k8sExecution.js]: Sounds good~ moved to a separate `createDbPasswords` function

**Comment by t83714** [comment]
@AlexGilleran 
Good idea~
The namespace currently can be overriden by two env variables:
- $CI_COMMIT_REF_SLUG
- $CLUSTER_NAMESPACE

It's a bit redundant.

How about I change it to:
```
Specify a namespace or prefill by env variable in future?
```
--- put default to `default` namespace seems a little bit scary
This way the default value is actually blank string.
If no env variable present at runtime, the program will report an error and exit.

**Comment by AlexGilleran** [comment]
Sounds good 👍

**Comment by t83714** [comment]
@AlexGilleran 
Has updated the question logic accordingly~
Please have another look when you get a chance~ 😄

**Pull request merged**